File: backend/app/schemas/monitor.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class GpuMetricPoint(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    timestamp: datetime
    node_id: str
    gpu_index: int
    utilization: int
    memory_used: int
    memory_total: int
    temperature: int
    power_usage: int


class NodeOverview(BaseModel):
    node_id: str
    hostname: str | None = None
    ip: str
    status: str
    gpu_count: int
    gpu_used: int
    cpu_percent: float
    memory_percent: float
    last_heartbeat: datetime | None = None

File: frontend/src/api/monitor.ts
import request from './request'

export interface GpuMetricPoint {
  timestamp: string;
  node_id: string;
  gpu_index: number;
  utilization: number;
  memory_used: number;
  memory_total: number;
  temperature: number;
  power_usage: number;
}

export interface NodeOverview {
  node_id: string;
  hostname?: string | null;
  ip: string;
  status: string;
  gpu_count: number;
  gpu_used: number;
  cpu_percent: number;
  memory_percent: number;
  last_heartbeat?: string | null
}

export async function fetchNodesOverview() {
  return request.get<unknown, NodeOverview[]>('/monitor/nodes/overview')
}

export async function fetchGpuHistory(nodeId: string, hours = 1) {
  return request.get<unknown, GpuMetricPoint[]>('/monitor/gpu/history', {
    params: { node_id: nodeId, hours },
  })
}
